crypto/src/dex/trading_pair.rs

```diff
@@ -1,17 +1,34 @@
+use anyhow::{anyhow, Result};
 use decaf377::FieldExt;
 use penumbra_proto::{dex as pb, Protobuf};
 
 use crate::asset;
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Copy)]
 pub struct TradingPair {
-    pub asset_1: asset::Id,
-    pub asset_2: asset::Id,
+    pub(crate) asset_1: asset::Id,
+    pub(crate) asset_2: asset::Id,
 }
 
 impl TradingPair {
+    pub fn new(asset_1: asset::Id, asset_2: asset::Id) -> Result<Self> {
+        if asset_2 < asset_1 {
+            return Err(anyhow!("asset_2 must be greater than asset_1"));
+        }
+
+        Ok(Self { asset_1, asset_2 })
+    }
```

Comment:
It'd be handy to have a constructor that picks the correct ordering -- the test code for the LPNFTs constructs a `TradingPair`, for instance, and without having a way to get the ordering right, it would fail the test on deserialization if the ordering happens to be wrong.

Reply:
Also, it looks like the ordering check is only applied in `new`, rather than also being applied when we deserialize the proto, so we should add it there too.

Reply:
@hdevalence what behavior should we have for `pub fn canonical_order_for(pair: (asset::Id, asset::Id)) -> Self;` if the same asset ID is passed twice? It seems like we might want to have a `Result` type returned, because an asset pair should consist of unique assets? Or should we just return a `TradingPair` with duplicated assets?